### DISCUSSION TITLE
Docker Compose YAML Invalid but Works Locally

### DISCUSSION THREAD

**Discussion by RickRen7575** [community/community#39333] [Actions] [answered]
### Select Topic Area

Bug

### Body

I'm having issues building my docker compose application in GitHub Actions because it is throwing errors that my compose YAML file is not formatted correctly. It works locally, and I just tested on a fresh Linux machine and it also worked, so I don't know what is going on. GitHub Actions is solely running a shell script that looks like this:

```
#! /usr/bin/env sh

# Exit in case of error
set -e

TAG=${TAG?Variable not set} \
FRONTEND_ENV=${FRONTEND_ENV-production} \
docker-compose \
-f ./docker-compose.yml \
-f ./docker-compose.build.yml \
build
```

And my error is the following:

```
the compose file '././docker-compose.build.yml' is invalid because: secrets.github_access_token value 'environment' does not match any of the regexes: '^x-' services.backend.build contains unsupported option: 'secrets'
```

For reference, here is my docker-compose.build.yml file:

```
version: '3.9'
services:
  backend:
    build:
      secrets:
        - GITHUB_ACCESS_TOKEN
secrets:
  GITHUB_ACCESS_TOKEN:
    environment: GITHUB_ACCESS_TOKEN
```

It's basically telling me that this format is not part of the spec, but I can clearly see it is based on documentation (and because it works locally and on the fresh linux machine). Here is a link to the corresponding documentation:

1. [Top level secrets spec](https://docs.docker.com/compose/compose-file/#secrets-top-level-element)
2. [Secrets element under build element](https://docs.docker.com/compose/compose-file/build/#secrets)

My docker compose version is 2.12.2 and my docker version is 20.10.21. My Linux distro is Ubuntu 22.10

**Comment by RickRen7575** [comment]
Also here is my output from running `docker info` on my machine:

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.9.1-docker)
  compose: Docker Compose (Docker Inc., v2.12.2)
  scan: Docker Scan (Docker Inc., v0.21.0)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 3
 Server Version: 20.10.21
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc io.containerd.runc.v2 io.containerd.runtime.v1.linux
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 1c90a442489720eec95342e1789ee8a5e1b9536f
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.19.0-23-generic
 Operating System: Ubuntu 22.10
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 1.929GiB
 Name: docker-compose-test
 ID: QJZW:55RB:YX7P:YRTY:P4E2:YMGH:OBLX:FFER:QE3C:FEX3:HVZR:R2FH
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by RickRen7575** [comment]
Also, I just tested using a similar strategy just building docker images explicitly instead of using `docker compose build` syntax through this: https://github.com/marketplace/actions/build-and-push-docker-images. In this syntax, I am able to pass build secrets securely, however, I'd like to be able to use `docker compose build` as that is a cleaner way to do this.

Not sure if that helps understand the problem at all...

**Comment by airtower-luna** [comment, marked as answer]
`docker-compose` (used in your script) and `docker compose` (which you mentioned later) are very different tools, the latter being meant as a replacement for the former. So I'm not surprised `docker-compose` doesn't support some newer features.

According to the software lists in the [actions/runner-images repository](https://github.com/actions/runner-images) `docker compose` should be installed on the runners, too (at least on `ubuntu-22.04` and `ubuntu-20.04`), so try using it in your script. :slightly_smiling_face:

**Reply by RickRen7575** [reply]
I get that but at the same time - why would this work locally with `docker-compose`? I'll try that though

**Reply by airtower-luna** [reply]
Possibly a different `docker-compose` version? You can check the software lists for the runners if you want to compare the versions, "Docker Compose v1" should be `docker-compose`, and "Docker Compose v2" `docker compose`. :sweat_smile:

**Reply by RickRen7575** [reply]
I got you okay that makes sense. Locally, `docker-compose version` was the same as `docker compose version` so I just assumed they would be in automation which was a bad assumption to make!

**Reply by RickRen7575** [reply]
That worked! Thank you very much @airtower-luna
